### PR TITLE
add arm, ppc, s390x to github build matrix (fix #16)

### LINKED ISSUE #16: Add arm, ppc, and s390x builds
If builds for s390x and ppc are added, the pipx install commands in pyproject.toml may need `-f` added to force install the correct versions of CMake and Ninja -- or a check to ensure the versions installed meet a minimum requirement.

Basically what this needs is:
- expand the build matrix in https://github.com/nightlark/swig-pypi/blob/main/.github/workflows/build.yml to include arm, ppc, and s390x build entries
- trigger the build workflow to confirm it works; fix any minor issue that may appear

### PR DESCRIPTION
**Pull request by frankschmitt**
First stab at #16 . I expect this to fail spectacularly in CI :-)

### CODE REVIEW

**Comment by nightlark** [comment]
Ah yes, the invalid workflow file is unrelated -- I just pushed a fix to it that you can merge into your branch (from the GitHub web UI there should be a  an option on https://github.com/frankschmitt/swig-pypi/tree/add-arm-ppc-s390x-builds-%2316 just below the green `Code` button that says `Fetch upstream`).

As a side note, if you don't want to wait for me to approve running the workflows for the PR, you can trigger it on your fork under the `Actions` tab (https://github.com/frankschmitt/swig-pypi/actions).

After turning on Actions in your fork, you can manually trigger a workflow run for the CI workflow from that same tab, just select the branch with your changes in it.

trigger-workflow-run

If a message about permissions comes up, you can change the settings for the repository to either `Allow all actions`, or `Allow select actions` (and then make sure `pypa/*` is on the list and `Actions created by GitHub` is selected).

actions-permissions

**Comment by frankschmitt** [comment]
Ok, I figured it out - after adding a new workflow and creating a PR to my own repo, the existing workflows magically appeared.

**Comment by frankschmitt** [comment]
> I think to resolve that pyproject.toml file can be updated to run `pipx install -f <dependency>` to force the install -- some of the docker images used by cibuildwheel have newer versions of cmake installed than others.
> 
> Maybe also bump up the cmake and ninja versions used at the same time to cmake==3.21.4 and ninja==1.10.2.2

Thanks, that did the trick. The build was successful; however, it generated 13 wheel files instead of 11 (apparently because for both Ubuntu builds, CI built manylinux_ and musllinux_ artifacts). I tried fixing that by adding if quemu .. conditions to set the environment variables CIBW... only when required (like scikit-build does), but this caused an error. I'll take another look tomorrow.

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -73,10 +73,11 @@ jobs:
         
       - name: Build wheels
         uses: pypa/cibuildwheel@v2.2.2
+        if: (!matrix.use_qemu) || fromJSON(env.USE_QEMU)
```
Comment (nightlark) [inline .github/workflows/build.yml]: It looks like you noticed the change here had unintended effects, more or less what happened is the `if` key applies to the entire step and is position independent within the step -- it's not like other programming languages where it just applies to subsequent lines of code.

```
      - if: (!matrix.use_qemu) || fromJSON(env.USE_QEMU)
        name: Build wheels
        uses: pypa/cibuildwheel@v2.2.2
        env:
          CIBW_ARCHS: "${{ matrix.arch }}"
          CIBW_BUILD: "cp39-${{ matrix.build }}*"
```
and
```
      - name: Build wheels
        uses: pypa/cibuildwheel@v2.2.2
        env:
          CIBW_ARCHS: "${{ matrix.arch }}"
          CIBW_BUILD: "cp39-${{ matrix.build }}*"
        if: (!matrix.use_qemu) || fromJSON(env.USE_QEMU)
```
 would all be equivalent to what you had and skip over the entire step if the condition isn't true.

---

And some info on musllinux/manylinux platform tags:

Building both musllinux and manylinux variants of the wheels for all platforms is fine; those platform tags are unrelated to the CPU (aarch64, s390x, etc) platform tags, and are used to indicate which libc implementation the wheel was built for (musl or glibc).

**Comment by nightlark** [comment]
> I tried fixing that by adding if quemu .. conditions to set the environment variables CIBW... only when required (like scikit-build does), but this caused an error. I'll take another look tomorrow.

I'd ignore the condition scikit-build is using on the step -- they have the condition on the step for a different purpose (limiting the number of slower running for builds running during a push/commit).

The place to change to only build "manylinux_" wheels for x86_64 and i686 wheels on Ubuntu would be changing the build matrix entries for those two jobs from `build: ""` to `build: "manylinux_"` -- though I think building the musllinux wheels there should be fine (those build still finish pretty fast anyway).

**Comment by frankschmitt** [comment]
If building both manylinux_ and musllinux_ on Ubuntu is fine, I'll keep it as it is. 
I've fixed the trailing whitespaces in build.yml; apart from that, are there any changes that still need to be done?

**Review by nightlark — APPROVED** [review]
Yea, the way manylinux and musllinux variants are being built in the same job on Ubuntu is fine.

Thanks for all your work on getting these builds running, looks good to me! :smile:

**Comment by frankschmitt** [comment]
@nightlark You’re welcome. To be completely honest, to me, it feels like you did most of the work anyway - thanks a ton for your patience and help. Highly appreciated!

**Pull request merged**